// db/group.js
var util = require('util'),
    Document = require('./document.js'),
    Collection = require('./collection.js'),
    CONST = require('./const.js'),
    VERSION = CONST.VERSION;

module.exports = Group;

/* public apis */
function Group (uuid, name, description, owner_ids) {

  this.uuid = uuid;
  this.name = name;
  this.description = description;
  this.owner_ids = owner_ids;
  this.version = VERSION;

  Document.call(this, Group._collectionName, Group._transformDBResult);
  return this;
}
util.inherits(Group, Document);

/* static properties */
Group._collectionName = 'groups';
Group._transformDBResult = function (result) {
  var ret;

  ret  = new Group(result.uuid, result.name, result.description, result.owner_ids);
  ret._copyDBResult(result);

  return ret;
};
Group.getCollection = Collection.getCollection.bind(null, Group._collectionName);
Group.findForEach = Collection.findForEach.bind(null, Group._collectionName, Group._transformDBResult);
Group.findOne = Collection.findOne.bind(null, Group._collectionName, Group._transformDBResult);


/* overwrite the _validate function */
Group.prototype._validate = function () {
  if (typeof this.uuid === 'string' &&
      typeof this.name === 'string' &&
      typeof this.description === 'string' &&
      Array.isArray(this.owner_ids)) {
    return true;
  } else {
    return false;
  }
};
